File: rust/src/bin/day7.rs
// ! FAILING

use aoc2023::read_file;
use std::collections::HashMap;

fn main() {
    let inn = read_file();

    // key_mapping
    let mut key_p1: HashMap<&str, i32> =
        HashMap::from_iter([("A", 14), ("K", 13), ("Q", 12), ("J", 11), ("T", 10)]);

    let mut key_p2: HashMap<&str, i32> =
        HashMap::from_iter([("A", 14), ("K", 13), ("Q", 12), ("J", 1), ("T", 10)]);

    key_p1.insert("2", 2);
    key_p1.insert("3", 3);
    key_p1.insert("4", 4);
    key_p1.insert("5", 5);
    key_p1.insert("6", 6);
    key_p1.insert("7", 7);
    key_p1.insert("8", 8);
    key_p1.insert("9", 9);

    key_p2.insert("2", 2);
    key_p2.insert("3", 3);
    key_p2.insert("4", 4);
    key_p2.insert("5", 5);
    key_p2.insert("6", 6);
    key_p2.insert("7", 7);
    key_p2.insert("8", 8);
    key_p2.insert("9", 9);

    // list tables
    let mut tab_p1: HashMap<&str, Vec<String>> = HashMap::new();
    let mut tab_p2: HashMap<&str, Vec<String>> = HashMap::new();

    let combinations = [
        "fiveok", "fourok", "fullh", "threeok", "twopair", "onepair", "hc",
    ];
    for key in combinations {
        tab_p1.insert(key, vec![]);
        tab_p2.insert(key, vec![]);
    }

    // bid table
    let mut bid_map: HashMap<&str, i32> = HashMap::new();

    for line in inn.lines() {
        let (hand, bid) = line.split_once(' ').unwrap();
        bid_map.insert(hand, bid.parse().unwrap());

        #[allow(non_snake_case)]
        let mut C: HashMap<char, usize> = HashMap::new();
        for c in hand.chars() {
            *C.entry(c).or_insert(0) += 1;
        }

        let p1_range: Vec<usize> = C.values().map(|x| x.clone()).collect();
        let mut p2_range: Vec<usize> = Vec::new();

        let mut maxx: usize = 0;
        let mut mididx: usize = 0;
        let mut jkr_cnt: usize = 0;
        for (idx, (key, v)) in C.iter().enumerate() {
            if *key == 'J' {
                jkr_cnt += 1;
                continue;
            }

            if maxx < *v {
                maxx = *v;
                mididx = idx;
            }

            p2_range.push(*v);
        }

        // Par 2 Relevant:  if all jokers
        if let Some(v) = p2_range.get_mut(mididx) {
            *v += jkr_cnt;
        } else {
            tab_p2.get_mut("fiveok").unwrap().push(hand.to_string());
            continue;
        }

        if p1_range.contains(&5) {
            tab_p1.get_mut("fiveok").unwrap().push(hand.to_string());
        } else if p1_range.contains(&4) {
            tab_p1.get_mut("fourok").unwrap().push(hand.to_string());
        } else if p1_range.contains(&3) && p1_range.contains(&2) {
            tab_p1.get_mut("fullh").unwrap().push(hand.to_string());
        } else if p1_range.contains(&3) {
            tab_p1.get_mut("threeok").unwrap().push(hand.to_string());
        } else if &p1_range.iter().filter(|x| x == &&2).count() == &2 {
            tab_p1.get_mut("twopair").unwrap().push(hand.to_string());
        } else if p1_range.contains(&2) {
            tab_p1.get_mut("onepair").unwrap().push(hand.to_string());
        } else {
            tab_p1.get_mut("hc").unwrap().push(hand.to_string());
        }

        if p2_range.contains(&5) {
            tab_p2.get_mut("fiveok").unwrap().push(hand.to_string());
        } else if p2_range.contains(&4) {
            tab_p2.get_mut("fourok").unwrap().push(hand.to_string());
        } else if p2_range.contains(&3) && p2_range.contains(&2) {
            tab_p2.get_mut("fullh").unwrap().push(hand.to_string());
        } else if p2_range.contains(&3) {
            tab_p2.get_mut("threeok").unwrap().push(hand.to_string());
        } else if &p2_range.iter().filter(|x| x == &&2).count() == &2 {
            tab_p2.get_mut("twopair").unwrap().push(hand.to_string());
        } else if p2_range.contains(&2) {
            tab_p2.get_mut("onepair").unwrap().push(hand.to_string());
        } else {
            tab_p2.get_mut("hc").unwrap().push(hand.to_string());
        }
    }

    let mut p1: i32 = 0;
    let mut p2: i32 = 0;
    for comb in combinations.iter().rev() {
        let p1_srted = merge_sort(tab_p1.get(*comb).unwrap().to_owned(), &key_p1);
        let p2_srted = merge_sort(tab_p2.get(*comb).unwrap().to_owned(), &key_p2);
        println!("{:?}", p1_srted);

        for (rank, hand) in p1_srted.iter().enumerate() {
            p1 += (rank + 1) as i32 * bid_map.get(hand.as_str()).unwrap();
        }

        for (rank, hand) in p2_srted.iter().enumerate() {
            p2 += (rank + 1) as i32 * bid_map.get(hand.as_str()).unwrap();
        }
    }

    println!("Part 1: {}", p1);
    println!("Part 2: {}", p2);
}

fn merge_sort(lst: Vec<String>, key: &HashMap<&str, i32>) -> Vec<String> {
    let len = lst.len();

    if len < 2 {
        return lst;
    }

    let mid = len / 2;

    let mut lpart: Vec<String> = Vec::new();
    let mut rpart: Vec<String> = Vec::new();

    for (idx, v) in lst.iter().enumerate() {
        if idx < mid {
            lpart.push(v.clone());
        } else {
            rpart.push(v.clone());
        }
    }

    println!("{:?}", lpart);
    println!("{:?}", rpart);

    let left = merge_sort(lpart, key);
    let right = merge_sort(rpart, key);

    let mut temp: Vec<String> = Vec::new();
    let (mut i, mut j) = (0, 0);
    while i < left.len() && j < right.len() {
        let lhand = left[i].as_str();
        let rhand = right[i].as_str();

        let lpiped: Vec<i32> = lhand
            .chars()
            .map(|x| *key.get(x.to_string().as_str()).unwrap())
            .collect();
        let rpiped: Vec<i32> = rhand
            .chars()
            .map(|x| *key.get(x.to_string().as_str()).unwrap())
            .collect();

        for (lv, rv) in lpiped.iter().zip(rpiped.iter()) {
            if lv < rv {
                temp.push(lhand.to_string());
                i += 1;
                break;
            }

            if rv < lv {
                temp.push(rhand.to_string());
                j += 1;
                break;
            }
        }
    }

    while i < left.len() {
        temp.push(left[i].clone());
        i += 1;
    }

    while j < right.len() {
        temp.push(right[j].clone());
        j += 1;
    }

    return temp;
}

File: rust/src/bin/day7.py
from collections import Counter, defaultdict

def ms(lst, tab):
    LEN = len(lst)
    if LEN < 2:
        return lst
    
    mid = LEN // 2
    
    left = ms(lst[:mid], tab)
    right = ms(lst[mid:], tab)


    temp = list()
    i, j = 0, 0
    while i < len(left) and j < len(right):
        lval = left[i]
        rval = right[j]

        lval = [tab[c] for c in lval]
        rval = [tab[c] for c in rval]

        for l, r in zip(lval, rval):
            if l < r:
                temp.append(left[i])
                i += 1
                break
            if r < l:
                temp.append(right[j])
                j += 1
                break


    while i < len(left):
        temp.append(left[i])
        i += 1
        
    while j < len(right):
        temp.append(right[j])
        j += 1

    return temp

with open('inputs/day7.in', 'r') as f:
    I = f.read().split("\n")

# cards strength
V = dict(A = 14, K = 13, Q = 12, J = 11, T = 10)
V2 = dict(A = 14, K = 13, Q = 12, J = 1, T = 10)
for i in range(2, 10):
    V[str(i)] = i
    V2[str(i)] = i

combinations = ['fiveok', 'fourok', 'fullh', 'threeok', 'twopair', 'onepair', 'hc']
S = defaultdict(list)
S2 = defaultdict(list)
bid_map = dict()

# part 1
for lines in I:
    if lines == '':
        continue

    [hand, bid] = lines.split(" ")
    bid_map[hand] = int(bid)
    C = Counter(hand)
    vals = C.values()

    if 5 in vals:
        S['fiveok'].append(hand)
    elif 4 in vals:
        S['fourok'].append(hand)
    elif 3 in vals and 2 in vals:
        S['fullh'].append(hand)
    elif 3 in vals:
        S['threeok'].append(hand)
    elif len([x for x in vals if x == 2]) == 2:
        S['twopair'].append(hand)
    elif 2 in vals:
        S['onepair'].append(hand)
    else:
        S['hc'].append(hand)


M = list()
for k in combinations[::-1]:
    srted = ms(S[k], V)
    M.extend(srted)

p1 = 0
for (rank, hand) in enumerate(M, 1):
    p1 += rank * bid_map[hand]

print(f'Part 1: {p1}')

# part 2
for lines in I:
    if lines == '':
        continue

    [hand, bid] = lines.split(" ")
    bid_map[hand] = int(bid)
    C = Counter(hand)

    vals = list()
    maxx = 0
    maxx_idx = 0 
    idx = 0
    jkr_cnt = 0
    for [k , v] in C.items():
        if k == 'J':
            jkr_cnt += v
            continue
        
        if maxx < v:
            maxx = v
            maxx_idx = idx
            

        vals.append(v)
        idx += 1

    if jkr_cnt == 5:
        S2['fiveok'].append(hand)
        continue

    vals[maxx_idx] += jkr_cnt

    if 5 in vals:
        S2['fiveok'].append(hand)
    elif 4 in vals:
        S2['fourok'].append(hand)
    elif 3 in vals and 2 in vals:
        S2['fullh'].append(hand)
    elif 3 in vals:
        S2['threeok'].append(hand)
    elif len([x for x in vals if x == 2]) == 2:
        S2['twopair'].append(hand)
    elif 2 in vals:
        S2['onepair'].append(hand)
    else:
        S2['hc'].append(hand)


M = list()
for k in combinations[::-1]:
    srted = ms(S2[k], V2)
    M.extend(srted)

p2 = 0
for (rank, hand) in enumerate(M, 1):
    p2 += rank * bid_map[hand]

print(f'Part 2: {p2}')